### ISSUE TITLE
Error when downloading biome data 

### ISSUE THREAD

**Issue by vuillota** [mapme-initiative/mapme.biodiversity#196]
When I try to download biome indicator from TEOW data, I face an error.

Here is a reproducible example.

```
#Import relevant packages
install.packages("mapme.biodiversity", "geodata", "sf")
library(geodata)
library(sf)
library(mapme.biodiversity)

#Download the multipolygon of Comoros and transform it to POLYGON, compatible with init_portfolio() function.
poly = gadm(country = "COM", resolution = 1, level = 0, path = paste(tempdir(), "matching_pre", sep = "/")) %>%
    st_as_sf() %>%
    st_make_valid() %>%
    st_cast("POLYGON")

#Create the portfolio, download TEOW data and compute the biome indicator
  data.bio = init_portfolio(poly,
                            years = 2000:2021,
                            outdir = paste(tempdir(), "matching_pre", sep = "/"),
                            add_resources = FALSE) %>%
    get_resources("teow") %>%
    calc_indicators(indicators = "biome") %>%
    unnest(biome)
```

The error is the following.

```
Error in `[<-`:
! Assigned data `results[indicator]` must be compatible with existing data.
 Existing data has 16 rows.
 Assigned data has 12 rows.
 Only vectors of size 1 are recycled.
Caused by error in `vectbl_recycle_rhs_rows()`:
! Can't recycle input of size 12 to size 16.
Run `rlang::last_trace()` to see where the error occurred.
```

From what I understand, the indicator is computed for some polygons in the portfolio (12/16) but not for others (4/16), leading to an error when constructing the dataframe. Also, it seems to me that the data are well downloaded (get_resources() step), and that the issue comes from the computation of the indicator (calc_indicators()). Note that I don't have such error with the other data and indicators I use.

I was wondering if you have faced similar issues, or had any suggestion. This could be very helpful !

**Comment by vuillota** [comment]
I dug into the functions used by calc_indicators() and was able to identify the source of the issue. 

~
TLDR : the issue seems to come from missing values in the TEOW data, leading to an error in the  ` .get_single_indicators()` component of the `calc_indicators()` function. Below, more details and a reproducible example of the issue, plus a modified version of ` .get_single_indicators()`. 
~

It comes from the ` .get_single_indicators` function. The biome indicator is computed first [here](https://github.com/mapme-initiative/mapme.biodiversity/blob/6f056e2dcb4fda829cffb20951d0a0d6cd689d2e/R/calc_indicators.R#L71), and the variable `results` is first a list of 16 lists (one for each polygon in the portfolio here) containing the indicator. Unexpectedly, 4/16 have a list with `character(0)` and `double(0)` for biomes and area variables. These lines are removed in [the next operation](https://github.com/mapme-initiative/mapme.biodiversity/blob/6f056e2dcb4fda829cffb20951d0a0d6cd689d2e/R/calc_indicators.R#L82) using ` mapme.biodiversity:::.bind_assets()` function. This explains the error faced at [this step](https://github.com/mapme-initiative/mapme.biodiversity/blob/6f056e2dcb4fda829cffb20951d0a0d6cd689d2e/R/calc_indicators.R#L86). _4 have also NA values, but these are not removed._

My guess is that we have no value in the TEOW data for these 4/16 polygons. From `.get_single_indicators` it is then possible to build a slightly modified function to add the biome variable with `left_join()` instead of `x[indicator] <- results[indicator]`, and avoid the error.

```
##INPUTS
### x : the portfolio where TEOW data have been downloaded through mapme.biodiversity::get_resources() function
### indicator : the indicator we want to compute from TEOW data ("biome" here).
fn_calc_biome_temp = function(x, indicator)
{
  i <- NULL
  args <- list() #No additionnal arguments
  atts <- attributes(x) #Attributes of the portfolio
  selected_indicator <- available_indicators(indicator) #The indicator available, processing mode and function to extract it
  processing_mode <- selected_indicator[[indicator]]$processing_mode #Get processing mode
  params <- mapme.biodiversity:::.check_resource_arguments(selected_indicator, 
                                                           args) #Check we do have the resource to compute the indicator
  params$verbose <- atts$verbose #Get verbose
  fun <- selected_indicator[[indicator]]$fun #Get the function
  available_resources <- atts$resources #The resources available in the temporary folder of mapme
  required_resources <- selected_indicator[[indicator]]$resources #The resource required to compute the indicator specified
  if (processing_mode == "asset") {
    p <- with_progress(progressr::progressor(steps = nrow(x)))
    results <- furrr::future_map(1:nrow(x), function(i) {
      p()
      resources <- mapme.biodiversity:::.prep(x[i, ], atts$resources, required_resources)
      mapme.biodiversity:::.compute(x[i, ], resources, fun, params, i)
    }, .options = furrr::furrr_options(seed = TRUE))
  }
  else {
    resources <- .prep(x, atts$resources, required_resources)
    results <- .compute(x, resources, fun, params, 1)
  }
  biome <- mapme.biodiversity:::.bind_assets(results) %>%
    mutate(.id = as.numeric(.id))
  #results <- nest(results, `:=`(!!indicator, !.id)) #The nesting is useless for this indicator
  #x[indicator] <- results[indicator] #this line is turned off as we do not nest
  data.biome = left_join(x, biome, by = c("assetid" = ".id"))
  # x <- relocate(x, !!attributes(x)[["sf_column"]], .after = last_col())
  # x
  
  return(data.biome)
  
}

```

And we can compute the indicator with no error.

```
#Download the data using get_resources
 tmp_pre = paste(tempdir(), "matching_pre", sep = "/") #the path where TEOW data are downloaded using
data = init_portfolio(poly,
                            years = 2000:2021,
                            outdir = tmp_pre,
                            add_resources = FALSE) %>%
    get_resources("teow")
data.bio = fn_calc_biome_temp(x = data, indicator = "biome")
```

Thus in `data.bio` we do have the 4 missing rows that correspond to the lists with `character(0)` and `double(0)` . We can then easily check whether these NA values in `data.bio` (8 in total) correspond to missing values in TEOW.

``` 
#Import the raw data from temporary R memory
data.teow = st_read(paste(tmp_pre, "teow/wwf_terr_ecos.gpkg", sep = "/")) 

#Extract biome variable from TEOW dataset, by cropping data on one of the 16 polygons.
## Choose the asset whose extent will crop the TEOW : 1-3, 10-13 do give biome, while the other 8 give no value.
n = 1 
## Extract the data on this extent
data.teow.test = data.teow %>%
  st_as_sf() %>%
  st_make_valid() %>%
  st_crop(st_bbox(data.bio[n,]))
```

**Comment by Jo-Schie** [comment]
Hi @vuillota . Great work! Thank you very much for digging into it! I am foreseeing that we can implement your suggested changes in the next two weeks. Can you use your workaround until this is fixed?

**Comment by goergen95** [comment]
Hi! Thanks for digging into this problem and showing what is causing the differences in lengths. I would argue that the bug is actually with the `biome` indicator functions, because it is returning 0 length tibbles while the requirement is to return NA if there are no matches. But we also do not check for 0 length returns in `.calc_single_indicator()` which in turn could return a more informative error message. I'll probably include both, a fix for the `biome` function and a 0-length check to fix this issue.

**Comment by goergen95** [comment]
Hi, I now merged into main. Could you please confirm that this fixes your issue?

**Comment by vuillota** [comment]
Hello, yes that works perfectly. Thanks !
